File: src/app/model/CadenceMetric.ts
import { MetricService } from "../services/metric.service";
import { Metric, MetricType } from "./Metric";
import { PowerMeterData } from "./PowerMeterData";
import { PowerMeterServiceEvent } from "./PowerMeterServiceEvent";

export class CadenceMetric extends Metric {
    private handler = this.handlePowerMeterEvent.bind(this);
    private lastPowerMeterDatas: PowerMeterData[] = [];

    constructor(metricService: MetricService) {
        super(MetricType.CADENCE, 'Cadence', 'rpm', metricService);
    }


    startLogging(): void {
        this.metricService.getPowerMeterService().subscribeForPowerData(this.handler);
    }


    stopLogging(): void {
        this.metricService.getPowerMeterService().unsubscribeForPowerData(this.handler);
    }

    handlePowerMeterEvent(event: PowerMeterServiceEvent): void {
        if (!event.powermeterData.cumulativeCrankRevolutions) {
            return;
        }

        const lastDataWithLessCrankRevolutions = this.getLastPowerMeterDataWithLessCrankRevolutions(event.powermeterData.cumulativeCrankRevolutions);
        if (lastDataWithLessCrankRevolutions === null) {
            this.addValue(0, new Date(event.powermeterData.timestamp));
        } else {
            const diffRotations = event.powermeterData.diffCrankRotations(lastDataWithLessCrankRevolutions) || 0;
            const timediff = event.powermeterData.diffCrankTimeInSeconds(lastDataWithLessCrankRevolutions) || 1;
            const speed = diffRotations / timediff; // in m/s

            this.addValue(speed, event.powermeterData.timestamp);
        }
        this.lastPowerMeterDatas.push(event.powermeterData);
        if (this.lastPowerMeterDatas.length > 10) {
            this.lastPowerMeterDatas.shift();
        }
    }

    getLastPowerMeterDataWithLessCrankRevolutions(cumulativeCrankRevolutions: number): PowerMeterData | null {
        for (let i = this.lastPowerMeterDatas.length - 1; i >= 0; i--) {
            const data = this.lastPowerMeterDatas[i];
            if (data.cumulativeCrankRevolutions && data.cumulativeCrankRevolutions < cumulativeCrankRevolutions) {
                return data;
            }
        }
        return null;
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/CadenceMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnPowerData;

import java.util.ArrayList;
import java.util.List;

public class CadenceMetric extends Metric implements OnPowerData {
    private final List<PowerMeterData> lastPowerMeterDatas = new ArrayList<>();

    public CadenceMetric(IMetricService metricService) {
        super(MetricType.CADENCE, "Cadence", "rps", metricService, "rpm");
    }

    @Override
    public void startLogging() {
        metricService.getPowerMeterService().subscribeForPowerData(this);
    }

    @Override
    public void stopLogging() {
        metricService.getPowerMeterService().unsubscribeForPowerData(this);
    }

    @Override
    public void onPowerData(PowerMeterData data) {
        if (data.cumulativeCrankRevolutions == null || data.cumulativeCrankRevolutions == 0) {
            return;
        }

        PowerMeterData lastDataWithLessCrankRevolutions =
                getLastPowerMeterDataWithLessCrankRevolutions(data.cumulativeCrankRevolutions);
        if (lastDataWithLessCrankRevolutions == null) {
            addValue(0, data.timestamp);
        } else {
            Double diffRotationsObj = data.diffCrankRotations(lastDataWithLessCrankRevolutions);
            Double timeDiffObj = data.diffCrankTimeInSeconds(lastDataWithLessCrankRevolutions);
            double diffRotations = diffRotationsObj == null ? 0 : diffRotationsObj;
            double timeDiff = timeDiffObj == null || timeDiffObj == 0 ? 1 : timeDiffObj;
            addValue(diffRotations / timeDiff, data.timestamp);
        }
        lastPowerMeterDatas.add(data);
        if (lastPowerMeterDatas.size() > 10) {
            lastPowerMeterDatas.remove(0);
        }
    }

    public PowerMeterData getLastPowerMeterDataWithLessCrankRevolutions(long cumulativeCrankRevolutions) {
        for (int i = lastPowerMeterDatas.size() - 1; i >= 0; i--) {
            PowerMeterData data = lastPowerMeterDatas.get(i);
            if (data.cumulativeCrankRevolutions != null
                    && data.cumulativeCrankRevolutions != 0
                    && data.cumulativeCrankRevolutions < cumulativeCrankRevolutions) {
                return data;
            }
        }
        return null;
    }
}
